src/inlines.c

```diff
@@ -1398,38 +1558,55 @@ static int parse_inline(cmark_parser *parser, subject *subj, cmark_node *parent,
   case '[':
     advance(subj);
     new_inl = make_str(subj, subj->pos - 1, subj->pos - 1, cmark_chunk_literal("["));
-    push_bracket(subj, false, new_inl);
+    push_bracket(subj, LINK, new_inl);
     break;
   case ']':
     new_inl = handle_close_bracket(parser, subj);
     break;
   case '!':
-    advance(subj);
-    if (peek_char(subj) == '[' && peek_char_n(subj, 1) != '^') {
+    if (peek_char_n(subj, 1) == '[' && peek_char_n(subj, 2) != '^') {
+      advance(subj);
       advance(subj);
       new_inl = make_str(subj, subj->pos - 2, subj->pos - 1, cmark_chunk_literal("!["));
-      push_bracket(subj, true, new_inl);
-    } else {
-      new_inl = make_str(subj, subj->pos - 1, subj->pos - 1, cmark_chunk_literal("!"));
+      push_bracket(subj, IMAGE, new_inl);
     }
     break;
-  default:
```

Comment:
What happenned to the default case? Did you just push it outside the switch?

Reply:
Yes. I had to move the `try_extensions` call outside of this switch, so that it could be tried with characters that are part of the built-in grammar (specifically `^` and `[`). That's also why i modified the `!` and `^` branches to peek for their syntax instead of eagerly advancing the pointer.